pkg/datapath/ipsec/ipsec_linux.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// +build linux
+
+package ipsec
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/datapath/linux"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/vishvananda/netlink"
+
+	"github.com/sirupsen/logrus"
+)
+
+type ipSecKey struct {
+	Spi   int
+	ReqID int
+	Auth  *netlink.XfrmStateAlgo
+	Crypt *netlink.XfrmStateAlgo
+}
+
+// ipSecKeysGlobal is safe to read unlocked because the only writers are from
+// daemon init time before any readers will be online.
+var ipSecKeysGlobal = make(map[string]*ipSecKey)
+
+func getIPSecKeys(ip net.IP) *ipSecKey {
+	key, scoped := ipSecKeysGlobal[ip.String()]
+	if scoped == false {
+		key, _ = ipSecKeysGlobal[""]
+	}
+	return key
+}
+
+func ipSecNewState() *netlink.XfrmState {
+	state := netlink.XfrmState{
+		Mode:  netlink.XFRM_MODE_TUNNEL,
+		Proto: netlink.XFRM_PROTO_ESP,
+		ESN:   false,
```

Comment:
You could enable [ESN](https://tools.ietf.org/html/rfc4304) which allows us to avoid key rotation which is required to mitigate replay attacks.

Reply:
agree adding it to the todo list.